### ISSUE TITLE
AdditiveAttention score calculation

### ISSUE THREAD

**Issue by ichenjia** [tensorflow/tensorflow#39332]
Bahdanau's paper specifies that the score is calculated as:

`score(st,hi)=v⊤atanh(Wa[st;hi]) `

While the actualy implemention's score is calcuted as:
```
    return math_ops.reduce_sum(
        scale * math_ops.tanh(q_reshaped + k_reshaped), axis=-1)
```

The explaination:
```

 1. Reshape `query` and `value` into shapes `[batch_size, Tq, 1, dim]`
     and `[batch_size, 1, Tv, dim]` respectively.
  2. Calculate scores with shape `[batch_size, Tq, Tv]` as a non-linear
     sum: `scores = tf.reduce_sum(tf.tanh(query + value), axis=-1)`
  3. Use scores to calculate a distribution with shape
     `[batch_size, Tq, Tv]`: `distribution = tf.nn.softmax(scores)`.
  4. Use `distribution` to create a linear combination of `value` with
     shape `batch_size, Tq, dim]`:
     `return tf.matmul(distribution, value)`.
```

Why such a difference? 

The Tensorflow's attention tutorial is more consistent with the paper:
https://www.tensorflow.org/tutorials/text/nmt_with_attention#restore_the_latest_checkpoint_and_test

**Comment by gowthamkpr** [comment]
@ichenjia Are you talking about [this](https://www.tensorflow.org/addons/api_docs/python/tfa/seq2seq/BahdanauAttention). Can you please explain the issue in detail. Thanks!

**Comment by ichenjia** [comment]
> @ichenjia Are you talking about [this](https://www.tensorflow.org/addons/api_docs/python/tfa/seq2seq/BahdanauAttention). Can you please explain the issue in detail. Thanks!

No, I am not talking about the Add-on. I am talking about how Keras implemented the BahanauAttention layer at 

https://github.com/tensorflow/tensorflow/blob/v2.1.0/tensorflow/python/keras/layers/dense_attention.py

The score calculation seems off to me.

**Comment by qlzh727** [comment]
Adding @roumposg who is the original author of the dense_attention in keras.

**Comment by roumposg** [comment]
Thanks for your comment. The reason we did that is to be consistent with the previous implementation for RNN `tf.contrib.seq2seq.BahdanauAttention`. The main difference is the weight matrices W1 and W2 in the [scores](https://www.tensorflow.org/tutorials/text/nmt_with_attention#write_the_encoder_and_decoder_model), right? You can do that by including two Dense layers before the Attention layer.

**Comment by ichenjia** [comment]
The weight is one issue, which as you mentioned can be resolved easily with Dense layers. I have bigger issue with the way you calculate the score. 

The paper's formular is:

`score(st,hi)=v⊤ tanh(Wa[st;hi])`

As you can see, they are concating st and hi. But in  your implementation, you added the q and k together:

`math_ops.reduce_sum(
        scale * math_ops.tanh(q_reshaped + k_reshaped), axis=-1)`

Why do you sum instead of concat?

**Comment by roumposg** [comment]
We followed eq (4) from the [encoder and decoder model](https://www.tensorflow.org/tutorials/text/nmt_with_attention#write_the_encoder_and_decoder_model). As you say, in the initial paper by Bahdanau et al, they use a feed-forward neural network to calculate the scores. But we did not implement that method.

AFAIK, newer models, Transformer, BERT etc, all use the dot-product attention, so we did not do more work on other ways of calculating the scores.

**Comment by ichenjia** [comment]
you are right that the newer models all used dot product or scaled dot product. But since your documentation specifically mentioned it is Bahdanau's model, I thought you'd adhere to  the original score calculation. Maybe the solution is to point this out in the documentation?

**Comment by roumposg** [comment]
You are right, maybe we should not mention Bahdanau in the documentation, or we should point to a better reference, or both. Thanks for pointing it out.

**Comment by danilo-silva-ufsc** [comment]
Not sure if I should reuse the same issue, but I noticed a typo in the documentation, namely, `value` was written in place of `key` in the score calculation. The correct explanation would be:
```
  1. Reshape `query` and `key` into shapes `[batch_size, Tq, 1, dim]`
     and `[batch_size, 1, Tv, dim]` respectively.
  2. Calculate scores with shape `[batch_size, Tq, Tv]` as a non-linear
     sum: `scores = tf.reduce_sum(tf.tanh(query + key), axis=-1)`
```

**Comment by tensorflowbutler** [comment]
Hi There,

 This is a stale issue. As you are using an older version of tensorflow, we are checking to see if you still need help on this issue. Please test the issue with the latest TensorFlow (TF2.7 and tf-nightly). If the issue still persists with the newer versions of TF, please feel free to open it in [keras-team/keras](https://github.com/keras-team/keras/issues) repository by providing details about the issue and a standalone code to reproduce the issue. Thanks! 

 Please note that Keras development has moved to a separate Keras-team/keras repository to focus entirely on only Keras. Thanks!
